app/lib/domains.server.ts

```diff
@@ -0,0 +1,138 @@
+import * as dns from './dns.server';
+import { RecordStatus } from '@prisma/client';
+import { prisma } from '~/db.server';
+import type { RecordType, Record } from '@prisma/client';
+
+interface DnsRecord {
+  type: RecordType;
+  name: string;
+  value: string;
+}
+
+interface DomainRecord {
+  id?: Record['id'];
+  username?: Record['username'];
+  description?: Record['description'];
+  course?: Record['course'];
+  ports?: Record['ports'];
+}
+
+type Domain = DnsRecord & DomainRecord;
+
+export const createUserDomain = async (data: Domain) => {
+  if (!data.username) {
+    throw new Error('Username is not provided');
+  }
+
+  const existingRecords = await prisma.record.count({
+    where: {
+      name: data.name,
+      type: data.type,
+      value: data.value,
+    },
+  });
+
+  if (existingRecords > 0) {
+    throw new Error('Record already exists');
+  }
+
+  await dns.createRecord(data.type, data.name, data.value);
```

Comment:
You might as well run these two operations at the same time, since they don't depend on each other:

```ts
const [result,] = await Promise.all([
  // Add to database
  prisma.record.create({
    data: {
      username: data.username,
      type: data.type,
      name: data.name,
      value: data.value,
      expiresAt: setMonthsFromNow(6),
      status: RecordStatus.pending,
    },
  }),
  // Add to Route53
  dns.createRecord(data.type, data.name, data.value);
])
```

Question: can the `dns.createRecord()` throw?  If so, we should be careful about that too.

Reply:
It throws. In this case, is this how I should handle? 
```js
const [result, ] = Promise.all([p1.catch((error) => logger.warn(error)), p2.catch((error) => logger.warn(error))]);
```

Reply:
Yeah, for the `dns.createRecord()` add a `.catch()` like you did to log the problem.

The alternative is to not deal with any of these errors, and to just let them throw.  The caller would have to `try/catch`.

Reply:
I would like to go with the alternative. Otherwise, I have to remove many `throw` scenario tests.

Reply:
Sounds good.